jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-backend/src/main/java/org/jbpm/workbench/cm/backend/server/RemoteCaseManagementServiceImpl.java

```diff
@@ -159,5 +172,81 @@ public void removeComment(final String serverTemplateId, final String containerI
         return comparing(CaseMilestoneSummary::getStatus).thenComparing(request.getSortByAsc() ? comparatorByName: comparatorByName.reversed());
     }
 
+    @Override
+    public Actions getCaseActions(String serverTemplateId, String container, String caseId, String userId) {
+        Actions actions = new Actions();
+        actions.setAvailableActions(getAdHocActions(serverTemplateId, container, caseId));
+        actions.setInProgressAction(getInProgressActions(container, caseId));
+        actions.setCompleteActions(getCompletedActions(container, caseId));
+        return actions;
+    }
+
+    public List<CaseActionSummary> getInProgressActions(String containerId, String caseId) {
+        List<NodeInstance> activeNodes = client.getActiveNodes(containerId, caseId, 0, PAGE_SIZE_UNLIMITED);
+
+        return activeNodes.stream()
+                .map(new CaseActionNodeInstanceMapper())
+                .map(s -> {
+                    if (Arrays.asList(NODE_TYPE_HUMAN_TASK).contains(s.getType())) {
+                        s.setActualOwner(userTaskServicesClient.findTaskByWorkItemId(s.getId()).getActualOwner());
+                    }
+                    s.setActionType(CaseActionType.INPROGRESS);
+                    return s;
+                })
+                .collect(toList());
+    }
+
+    public List<NodeInstance> getCaseCompletedNodes(String containerId, String caseId) {
+        //TODO return client.getCompletedNodes(containerId, caseId, 0, PAGE_SIZE_UNLIMITED);
+        return new ArrayList<>();
+
+    }
+
+    public List<CaseActionSummary> getCompletedActions(String containerId, String caseId) {
+        List<NodeInstance> activeNodes = getCaseCompletedNodes(containerId, caseId);
+        return activeNodes.stream()
+                .map(new CaseActionNodeInstanceMapper())
+                .map(s -> {
+                    s.setActionType(CaseActionType.COMPLETED);
+                    return s;
+                })
+                .collect(toList());
+    }
+
+    public List<CaseActionSummary> getAdHocFragments(String containerId, String caseId) {
+        return client.getAdHocFragments(containerId, caseId)
+                .stream()
+                .map(new CaseActionAdHocMapper())
+                .collect(toList());
+    }
+
+    public List<CaseActionSummary> getAdHocActions(String serverTemplateId, String containerId, String caseId) {
+        List<CaseActionSummary> adHocActions = getAdHocFragments(containerId, caseId);
+        CaseInstanceSummary caseInstanceSummary = getCaseInstance(serverTemplateId, containerId, caseId);
+        List<CaseActionSummary> caseAdHocActions = caseInstanceSummary.getStages().stream()
+                .filter(s -> s.getStatus().equals(CaseStageStatus.ACTIVE.getStatus()))
+                .flatMap(ah -> ah.getAdHocActions().stream()).collect(toList());
```

Comment:
You can add the items directly to the final list, something like `.forEach(ah -> adHocActions.addAll(ah.getAdHocActions()));` That avoids the if statement and the extra caseAdHocActions collection.